lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,607 @@
+end
+
+function grid:after_draw_children(context, cr, width, height)
+    -- TODO support transparent colors
+    if self._private.border_width and self._private.border_width > 0 then
+        local bc = self._private.border_color
+
+        if bc then
+            cr:set_source(bc)
+        end
+
+        local bw = self._private.border_width
+        local max_col, max_row, mw, mh = get_ajusted_cell_sizes(
+            self, context, width, height
+        )
+        local spacing = (self._private.spacing or 0)
+            + (self._private.border_width or 0)
+
+        cr:set_line_width(bw)
+
+
+        local sw, sh = 0, 0
+
+        local max_x = math.min(width-bw , mw+#max_col*spacing)
+        local max_y = math.min(height-bw, mh+#max_col*spacing)
+
+        cr:rectangle( bw/2, bw/2, max_x, max_y )
+
+        for i=1, #max_row do
+            cr:move_to(bw/2, math.max(0, sh - spacing/2))
+            cr:line_to(max_x, math.max(0, sh - spacing/2))
+
+            sh = sh + max_row[i] + spacing
+        end
+
+        for i=1, #max_col do
+            cr:move_to(math.max(0, sw - spacing/2), bw/2)
+            cr:line_to(math.max(0, sw - spacing/2), max_y)
+
+            sw = sw + max_col[i] + spacing
+        end
+
+        cr:stroke()
+    end
+end
+
+local function get_next_empty(self, row, column)
+    row, column = row or 1, column or 1
+    local cc = self._private.column_count
+    for i = row, math.huge do
+        local r = self._private.rows[i]
+
+        if not r then
+            r = {}
+            self._private.rows[i] = r
+            self._private.row_count = i
+
+            return i, 1
+        end
+
+        for j = column, cc do
+            if not r[j] then
+                return i, j
+            end
+        end
+    end
+end
+
+local function index_to_point(data, index)
+    local counter = 0
+
+    -- As the widget matrix is sparse, the modulo cannot be used
+    for i=1, data._private.row_count do
+        for j=1, data._private.column_count do
+            counter = counter + (data._private.rows[i][j] and 1 or 0)
+
+            if counter == index then
+                return i, j
+            end
+        end
+    end
+
+    -- This can have bad side effects, but is the best fallback
+    return math.floor(index / data._private.row_count) + 1,
+        index % data._private.column_count + 1
+end
+
+--- Add some widgets to the given grid layout
+-- @param ... Widgets that should be added (must at least be one)
+function grid:add(...)
+
+    local args = { n=select('#', ...), ... }
+    assert(args.n > 0, "need at least one widget to add")
+
+    local row, column
+
+    for i=1, args.n do
+        local w = args[i]
+
+        row, column = get_next_empty(self, row, column)
+
+        base.check_widget(w)
+
+        self._private.rows[row][column] = w
+        table.insert(self._private.children, w)
+    end
+
+    self:emit_signal("widget::redraw_needed")
+    self:emit_signal("widget::layout_changed")
```

Comment:
Assuming `check_widget` raises an error, then these signals aren't emitted. I guess these signals either have to be emitted inside of the `for`-loop, or the `check_widget` is moved into a separate `for` loop that runs first before anything else is done.